crates/flipperzero/src/furi.rs

```diff
@@ -38,3 +40,60 @@ pub fn sleep(duration: core::time::Duration) {
         }
     }
 }
+
+/// MessageQueue provides a safe wrapper around the furi message queue primitive.
+pub struct MessageQueue<M: Sized> {
+    hnd: *const sys::message_queue::FuriMessageQueue,
+    _marker: core::marker::PhantomData<M>,
+}
+
+impl<M: Sized> MessageQueue<M> {
+    /// Constructs a message queue with the given capacity.
+    pub fn new(capacity: u32) -> Self {
+        Self {
+            hnd: unsafe { sys::message_queue::alloc(capacity, core::mem::size_of::<M>()) },
+            _marker: core::marker::PhantomData::<M>,
+        }
+    }
+
+    // Attempts to add the message to the end of the queue, waiting up to timeout ticks.
+    // FuriStatusOk (0) is returned if the insert is successful, otherwise FuriStatusErrorTimeout
+    // (-2) is returned.
+    pub fn put(&self, mut msg: M, timeout: u32) -> u32 {
```

Comment:
Can you use [`Duration`](https://doc.rust-lang.org/core/time/struct.Duration.html) as the type for `timeout`, rather than `u32`.

Reply:
`timeout` is actually an integer of ticks rather than an actual time. I'll change the name to `timeout_ticks` to avoid confusion.

Reply:
Oh interesting! Sorry for not seeing that in the comment.

For the high-level API, I think we would want to pass it through the equivalent of [`pdMS_TO_TICKS`](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/4c4089b1544b590ed3b72491a15365ec020c921f/include/projdefs.h#L42), so the user doesn't have to do this calculation themselves.

Reply:
Hmmm, do you know what the API would be for this in Furi?

If it turns out that the fz does frequency scaling to help manage its power budget, im not sure theres a good way to map to a duration without doing an API lookup for the frequency in the hot path.